tests/unit_and_functional/test_workflow_execution.py

```diff
@@ -168,6 +170,116 @@ def test_run_workflow_adds_git_commit_to_batch_and_service_env_vars(
     )
 
 
+@patch("bodywork.workflow_execution.rmtree")
+@patch("bodywork.workflow_execution.requests")
+@patch("bodywork.workflow_execution.download_project_code_from_repo")
+@patch("bodywork.workflow_execution.get_git_commit_hash")
+@patch("bodywork.workflow_execution.k8s")
+def test_run_workflow_runs_failure_stage_on_failure(
+    mock_k8s: MagicMock,
+    mock_git_hash: MagicMock,
+    mock_git_download: MagicMock,
+    mock_requests: MagicMock,
+    mock_rmtree: MagicMock,
+    project_repo_location: Path,
+):
+    config_path = Path(f"{project_repo_location}/bodywork.yaml")
+    config = BodyworkConfig(config_path)
+    config.project.run_on_failure = "on_fail_stage"
+
+    error_message = "Test Error"
+    mock_job = MagicMock(k8sclient.V1Job)
+    mock_k8s.configure_batch_stage_job.side_effect = [
+        k8sclient.ApiException(error_message),
+        mock_job,
+    ]
+    expected_result = [
+        k8sclient.V1EnvVar(name=FAILURE_EXCEPTION_K8S_ENV_VAR, value=error_message)
+    ]
+    mock_k8s.create_k8s_environment_variables.return_value = expected_result
+    mock_k8s.configure_env_vars_from_secrets.return_value = []
+
+    try:
+        run_workflow(config, "foo_bar_foo_993", project_repo_location)
+    except BodyworkWorkflowExecutionError:
+        pass
+
+    mock_k8s.configure_batch_stage_job.assert_called_with(
+        ANY,
+        "on_fail_stage",
+        ANY,
+        ANY,
+        ANY,
+        retries=ANY,
+        container_env_vars=expected_result,
+        image=ANY,
+        cpu_request=ANY,
+        memory_request=ANY,
+    )
+
+
+@patch("bodywork.workflow_execution.requests.Session")
+@patch("bodywork.workflow_execution.k8s")
+def test_failure_stage_does_not_run_for_docker_image_exception(
+    mock_k8s: MagicMock, mock_session: MagicMock, project_repo_location: Path
+):
+    config_path = Path(f"{project_repo_location}/bodywork.yaml")
+    config = BodyworkConfig(config_path)
+    response = requests.Response()
+    response.status_code = 401
+    mock_session().get.return_value = response
```

Comment:
`mock_session().get.return_value = requests.Response()` would be cleaner?

Reply:
I agree!